sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/ReplicatedResourceClient.java

```diff
@@ -158,37 +163,63 @@ public Mono<StoreResponse> invokeAsync(RxDocumentServiceRequest request,
-                            }
-                        }
-                    });
-            }
+        if (!(strategy instanceof ThresholdBasedAvailabilityStrategy)) {
+            // If no threshold based availability strategy is defined just return original Mono
+            return getStoreResponseMono(request, forceRefreshAndTimeout);
         }
 
-        // If the above conditions are not met, then we will just return the original request
-        if (monoList.isEmpty()) {
-            monoList.add(getStoreResponseMono(request, forceRefreshAndTimeout));
+        List<Mono<StoreResponse>> monoList = new ArrayList<>();
+        List<RxDocumentServiceRequest> requestList = new ArrayList<>();
+        final List<URI> effectiveEndpoints = getApplicableEndPoints(request);
+        if (effectiveEndpoints == null || effectiveEndpoints.stream().noneMatch(locationURI -> locationURI != null)) {
+            // If no endpoints are available just return original Mono
+            return getStoreResponseMono(request, forceRefreshAndTimeout);
         }
 
-        return Mono.firstWithValue(monoList);
+        effectiveEndpoints
+            .stream().filter(locationURI -> locationURI != null)
+            .forEach(locationURI -> {
+                RxDocumentServiceRequest newRequest = request.clone();
+                newRequest.requestContext.routeToLocation(locationURI);
+                requestList.add(newRequest);
+                if (monoList.isEmpty()) {
+                    monoList.add(getStoreResponseMono(newRequest, forceRefreshAndTimeout));
+                } else {
+                    monoList
+                        .add(
+                            getStoreResponseMono(newRequest, forceRefreshAndTimeout)
+                                .delaySubscription(((ThresholdBasedAvailabilityStrategy) strategy)
+                                    .getThreshold()
+                                    .plus(((ThresholdBasedAvailabilityStrategy) strategy)
+                                        .getThresholdStep()
+                                        .multipliedBy(monoList.size() - 1))
+                                ));
+                }
+            });
+
+        return Mono
+            .firstWithValue(monoList)
+            .onErrorMap(throwable -> {
+                Throwable exception = Exceptions.unwrap(throwable);
+
+                if (exception instanceof NoSuchElementException) {
+
+                    List<Throwable> innerThrowables = Exceptions
+                        .unwrapMultiple(exception.getCause());
+
+                    for (Throwable innerThrowable : innerThrowables) {
+                        Throwable innerException = Exceptions.unwrap(innerThrowable);
+
+                        // collect latest CosmosException instance bubbling up for a region
+                        if (innerException instanceof CosmosException) {
```

Comment:
what if the `innerException` is not instance of `CosmosException` - is it worth logging? - may be debug or warn since the request failed on the other region?

Reply:
Hmm - I don't mind adding a log because we would never expect this to happen. NoSuchElementException should happen when it is empty - otherwise we would always expect CosmosException - let me add a trace log if NoSucheElementException and warn otherwise

Reply:
Fixed